src/datacite/datacite.py

```diff
@@ -0,0 +1,311 @@
+        if resp.status_code == 200:
+            return resp
+        if resp.status_code in (429, 500, 502, 503, 504):
+            time.sleep(backoff)
+            backoff = min(backoff * 2, max_backoff)
+            continue
+        resp.raise_for_status()
+
+
+def _build_user_agent_headers(user_agent_mailto: Optional[str] = None) -> Dict[str, str]:
+    """Build HTTP headers with User-Agent for DataCite API.
+
+    DataCite provides higher rate limits (1000 requests per 5 minutes) for
+    identified unauthenticated users who include an email in the User-Agent.
+
+    Args:
+        user_agent_mailto: Email address to include in User-Agent header
+            for increased rate limits. If None, no User-Agent is set.
+
+    Returns:
+        Dictionary of HTTP headers. Empty if no email provided, otherwise
+        contains User-Agent header with format:
+        "SageBionetworks-DPE/1.0 (mailto:email@example.com)".
+    """
+    headers = {}
+    # Identified unauthenticated users get 1000 req/5min when User-Agent includes mailto.
+    # Example: "Sage-DPE/1.0 (mailto:phil.snyder@sagebase.org)"
+    if user_agent_mailto:
+        headers["User-Agent"] = f"SageBionetworks-DPE/1.0 (mailto:{user_agent_mailto})"
+    return headers
+
+
+def _should_continue_pagination(data: List[Dict[str, Any]], page_size: int) -> bool:
+    """Determine if pagination should continue based on returned data.
+
+    Pagination stops when either no data is returned or a short page
+    (fewer items than page_size) is received, indicating the last page.
+
+    Args:
+        data: List of data items returned from the current page.
+        page_size: Expected number of items per full page.
+
+    Returns:
+        True if pagination should continue (full page received),
+        False if this is the last page or no data was returned.
+    """
+    return bool(data) and len(data) >= page_size
+
+
+def _serialize_to_ndjson(obj: Dict[str, Any]) -> bytes:
+    """Serialize a single object to newline-delimited JSON bytes.
+
+    Converts a Python dictionary to a compact JSON string (no spaces),
+    encodes to UTF-8, and appends a newline character for NDJSON format.
+
+    Args:
+        obj: Dictionary object to serialize.
+
+    Returns:
+        UTF-8 encoded bytes containing JSON data followed by newline.
+    """
+    line = json.dumps(obj, separators=(",", ":"), ensure_ascii=False)
+    return line.encode("utf-8") + b"\n"
+
+
+def fetch_doi_page(
+    session: requests.Session,
+    prefixes: List[str],
+    state: str,
+    page_size: int,
+    page_number: int,
+    detail: bool,
+) -> Dict[str, Any]:
+    """Fetch a single page of DOI objects from DataCite API.
+
+    Args:
+        session: Requests session object with any required headers pre-configured.
+        prefixes: List of DataCite DOI prefixes to query (e.g., ["10.7303"]).
+        state: DOI state to filter by ("findable", "registered", or "draft").
```

Comment:
Are these the available states? Should we check / validate the input that it's one of these? 

Similarly, are there expected inputs formats for `prefixes`, `page_size`, `page_number`

Reply:
I thought that `state` would be enforced server-side, but since "findable" is the only DOI state that you can query if you're unauthenticated, it seems like it ignores whatever I provide for `state` and just returns "findable" DOIs.

I'll add a check.